Panic on drain capture meet unexpected min workload when try to the the target capture

### What did you do?

Enable scheduler v3 and run integration tests

### What did you expect to see?

No panic.

### What did you see instead?

Panic

```
panic: tpscheduler: drain capture meet unexpected min workload when try to the the target capture

goroutine 1095 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc004bbb2c0, {0x0, 0x0, 0x0})
	/go/pkg/mod/go.uber.org/zap@v1.21.0/zapcore/entry.go:232 +0x60a
go.uber.org/zap.(*Logger).Panic(0xc00066e3c0?, {0x6ccdf4d, 0x5a}, {0x0, 0x0, 0x0})
	/go/pkg/mod/go.uber.org/zap@v1.21.0/logger.go:230 +0x6e
github.com/pingcap/log.Panic({0x6ccdf4d, 0x5a}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/pingcap/log@v1.1.0/global.go:54 +0xe9
github.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*drainCaptureScheduler).Schedule(0xc0022a1950, 0xc002d80000?, {0xc004b0cfc0?, 0x4?, 0xc0032117c0?}, 0xc0022a17d0?, 0xc0022a1770?)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/scheduler_drain_capture.go:180 +0x1353
github.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*schedulerManager).Schedule(0xc0042483c0, 0xc001bc0930?, {0xc004b0cfc0, 0x4, 0x4}, 0x18bc41a?, 0x18978f7?, 0xc0022a17a0)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/scheduler_manager.go:74 +0x1fb
github.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*coordinator).poll(0xc0002b8510, {0x7f04beb10c40, 0xc000b1e2b8}, 0xc004a4c2a8?, {0xc004b0cfc0, 0x4, 0x4}, 0x1?)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/coordinator.go:254 +0x750
github.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*coordinator).Tick(0xc0002b8510, {0x7f04beb10c40, 0xc000b1e2b8}, 0x4?, {0xc004b0cfc0, 0x4, 0x4}, 0x0?)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/coordinator.go:108 +0x134
github.com/pingcap/tiflow/cdc/owner.(*changefeed).tick(0xc000e7b8c0, {0x81c8720?, 0xc000b1e2b8}, 0xc000ee38f0, 0xc004a4c8c8?)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:264 +0xe22
github.com/pingcap/tiflow/cdc/owner.(*changefeed).Tick(0xc000e7b8c0, {0x81c8630?, 0xc000b1e2a0}, 0x8176a58?, 0x1870395?)
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:160 +0x287
github.com/pingcap/tiflow/cdc/owner.(*ownerImpl).Tick(0xc000ee04d0, {0x7f04da109de8?, 0xc002eba080}, {0x819a0e8?, 0xc000d1bb40})
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/owner.go:203 +0xb25
github.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run(0xc002fa6980, {0x7f04da109de8, 0xc002eba080}, 0xc000ec4540, 0xbebc200, {0x6b9f109, 0x5})
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:247 +0x1344
github.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker(0xc0006d20d0, {0x81c86d0, 0xc002eba080}, {0x7f04d8743fe8, 0xc000ee04d0}, {0x819a0e8, 0xc000d1bb40}, 0x0?, {0x6b9f109, 0x5})
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:463 +0x1a5
github.com/pingcap/tiflow/cdc/capture.(*captureImpl).campaignOwner(0xc0006d20d0, {0x81c86d0?, 0xc00251c7a0})
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:436 +0xd52
github.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func2()
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:312 +0x133
created by github.com/pingcap/tiflow/cdc/capture.(*captureImpl).run
	/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:306 +0x6c6

[2022/07/08 14:18:27.334 +08:00] [INFO] [capture_manager.go:245] ["tpscheduler: all capture initialized"] [captureCount=2]
[2022/07/08 14:18:27.335 +08:00] [INFO] [replication_set.go:172] ["tpscheduler: initialize replication set"] [replicationSet="{\"TableID\":67,\"State\":\"Replicating\",\"Primary\":\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\",\"Secondary\":\"\",\"Captures\":{\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\":{}},\"Checkpoint\":{\"checkpoint_ts\":434441054476304385,\"resolved_ts\":434441055524618241}}"]
[2022/07/08 14:18:27.335 +08:00] [INFO] [replication_set.go:172] ["tpscheduler: initialize replication set"] [replicationSet="{\"TableID\":69,\"State\":\"Replicating\",\"Primary\":\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\",\"Secondary\":\"\",\"Captures\":{\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\":{}},\"Checkpoint\":{\"checkpoint_ts\":434441054476304385,\"resolved_ts\":434441055524618241}}"]
[2022/07/08 14:18:27.335 +08:00] [INFO] [replication_set.go:172] ["tpscheduler: initialize replication set"] [replicationSet="{\"TableID\":71,\"State\":\"Replicating\",\"Primary\":\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\",\"Secondary\":\"\",\"Captures\":{\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\":{}},\"Checkpoint\":{\"checkpoint_ts\":434441054476304385,\"resolved_ts\":434441055524618241}}"]
[2022/07/08 14:18:27.336 +08:00] [INFO] [replication_set.go:172] ["tpscheduler: initialize replication set"] [replicationSet="{\"TableID\":65,\"State\":\"Replicating\",\"Primary\":\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\",\"Secondary\":\"\",\"Captures\":{\"eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6\":{}},\"Checkpoint\":{\"checkpoint_ts\":434441054476304385,\"resolved_ts\":434441055524618241}}"]
[2022/07/08 14:18:27.338 +08:00] [INFO] [scheduler_drain_capture.go:101] ["tpscheduler: drain a stopping capture"] [captureID=eff86bac-63e4-4d2d-bf8b-13db0a4fb7e6]
[2022/07/08 14:18:27.339 +08:00] [PANIC] [scheduler_drain_capture.go:180] ["tpscheduler: drain capture meet unexpected min workload when try to the the target capture"] [stack="github.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*drainCaptureScheduler).Schedule\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/scheduler_drain_capture.go:180\ngithub.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*schedulerManager).Schedule\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/scheduler_manager.go:74\ngithub.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*coordinator).poll\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/coordinator.go:254\ngithub.com/pingcap/tiflow/cdc/scheduler/internal/v3.(*coordinator).Tick\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/scheduler/internal/v3/coordinator.go:108\ngithub.com/pingcap/tiflow/cdc/owner.(*changefeed).tick\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:264\ngithub.com/pingcap/tiflow/cdc/owner.(*changefeed).Tick\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:160\ngithub.com/pingcap/tiflow/cdc/owner.(*ownerImpl).Tick\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/owner.go:203\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:247\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:463\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).campaignOwner\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:436\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func2\n\t/home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:312"]
```

### Versions of the cluster

TiCDC version (execute `cdc version`):

```console
master
```

--- a/cdc/scheduler/internal/v3/scheduler_drain_capture.go
+++ b/cdc/scheduler/internal/v3/scheduler_drain_capture.go
@@ -102,16 +102,16 @@ func (d *drainCaptureScheduler) Schedule(
 			zap.String("captureID", stopping))
 	}
 
-	var availableCaptureCount int
+	captureWorkload := make(map[model.CaptureID]int)
 	for id := range captures {
 		if id != d.target {
-			availableCaptureCount++
+			captureWorkload[id] = 0
 		}
 	}
 
 	// this may happen when inject the target, there is at least 2 alive captures
 	// but when schedule the task, only owner alive.
-	if availableCaptureCount == 0 {
+	if len(captureWorkload) == 0 {
 		log.Warn("tpscheduler: drain capture scheduler ignore drain target capture, "+
 			"since cannot found destination captures",
 			zap.String("target", d.target), zap.Any("captures", captures))
@@ -121,7 +121,6 @@ func (d *drainCaptureScheduler) Schedule(
 
 	// victims record all table instance should be dropped from the target capture
 	victims := make([]model.TableID, 0)
-	captureWorkload := make(map[model.CaptureID]int)
 	for tableID, rep := range replications {
 		if rep.State != ReplicationSetStateReplicating {
 			// only drain the target capture if all tables is replicating,
@@ -138,7 +137,7 @@ func (d *drainCaptureScheduler) Schedule(
 
 		// only calculate workload of other captures not the drain target.
 		if rep.Primary != d.target {
-			captureWorkload[rep.Primary] += 1
+			captureWorkload[rep.Primary]++
 		}
 	}
 

--- a/cdc/scheduler/internal/v3/scheduler_drain_capture_test.go
+++ b/cdc/scheduler/internal/v3/scheduler_drain_capture_test.go
@@ -133,3 +133,23 @@ func TestDrainSkipOwner(t *testing.T) {
 	require.Len(t, tasks, 0)
 	require.EqualValues(t, captureIDNotDraining, scheduler.getTarget())
 }
+
+func TestDrainImbalanceCluster(t *testing.T) {
+	t.Parallel()
+
+	var checkpointTs model.Ts
+	currentTables := make([]model.TableID, 0)
+	captures := map[model.CaptureID]*CaptureStatus{
+		"a": {State: CaptureStateInitialized},
+		"b": {IsOwner: true, State: CaptureStateInitialized},
+	}
+	replications := map[model.TableID]*ReplicationSet{
+		1: {State: ReplicationSetStateReplicating, Primary: "a"},
+		2: {State: ReplicationSetStateReplicating, Primary: "a"},
+	}
+	scheduler := newDrainCaptureScheduler(10)
+	scheduler.setTarget("a")
+	tasks := scheduler.Schedule(checkpointTs, currentTables, captures, replications)
+	require.Len(t, tasks, 2)
+	require.EqualValues(t, "a", scheduler.getTarget())
+}

